wallet: allow duplicate open after auction expires

pinheadmz opened the pull request:
Closes #53 

This PR solves the following issue:

1.  hs-wallet send open `name`
2. generate enough blocks to expire auction without a single bid
3. re-attempt: hs-wallet send open `name`

This currently throws the error `Already sent an open for: name`, but is otherwise allowed by the protocol.

What I did was add one more check to `wallet.makeOpen()` so it won't throw if the name is expired.

Also adds a new test module `wallet-test-auction` which is a sort of integration test, because it needs to test an actual `WalletDB` (not `memwallet`) with a `chain` and `miner`.

pinheadmz commented:
Pulled out my attempt at a bug fix here after expanding the test a whole bunch. Gonna play with this some more...

pinheadmz commented:
OK, new approach here is to modify `txdb.isDoubleOpen(tx)` to ignore double opens if enough blocks have passed. This means that `txdb.add(tx, block)` and the `wallet.makeOpen()` check will allow a double open after enough blocks have passed.

That amount of time is `names.biddingPeriod + names.revealPeriod`, but reviewers should double-check me on that! If I pull out all the wallet checks and the txdb checks and the "will fail" tests, that value is enough blocks to pass the tests in `chain.verifyCovenants()`. Subtracting one block from that value will return a `bad-open-state` error from the chain.

boymanjor commented on test/wallet-test-auctions.js:
```diff
@@ -0,0 +1,184 @@
+/* eslint-env mocha */
```
Do you mind moving this to the `auctions-test.js` file?

pinheadmz replied:
The reason I didn't do this at first is because the current `auctions-test` seems to be testing consensus rules against `miner` and `chain`. That test uses `memwallet` which doesn't even [test for double opens](https://github.com/handshake-org/hsd/blob/master/lib/wallet/wallet.js#L1542-L1543).

Since this PR is focused on wallet behavior I thought it was best to start a new test file (other wallet auction functions can be added in here later). I could add `acutions-test` if you want, just wanted to throw out my motivation there...

boymanjor replied:
I see! Ok, can we call it `wallet-auctions-test.js` to stick to the current naming convention?

pinheadmz replied:
👍

boymanjor commented on test/wallet-test-auctions.js:
```diff
@@ -0,0 +1,184 @@
+  workers: workers
+});
+
+describe('Wallet Auction', function() {
```
I think that once we move this test to `auctions-test.js`, we should call it something like 'Test OPEN' and have it run after the 'Vickrey Auction' describe. What do you think?

boymanjor replied:
Resolved out of band.

pinheadmz commented:
@boymanjor I renamed the test file, then rebased on master. Turned out my test was interfering with other tests added since I first posted. So I added one more commit to play nice with others :-)
If that works for you, I'll squash.

turbomaze commented:
Seems to be working on testnet4! https://hnscan.com/name/chain

pinheadmz commented:
Rebased on master, just in case - didn't break any tests locally.

rsolari commented:
👍 Thanks! This change is an improvement on the current behavior. 

One note is that the new behavior will no longer block anyone from sending an `OPEN` for a `CLOSED` auction, even though a closed auction might have a winner. That seems ok, as long as there's some consensus rule that prevents that open from landing on the chain. The user just won't get feedback that their `OPEN` is invalid.

pinheadmz commented:
@rsolari were you able to OPEN a closed auction with a winner? That check should be happening already, out of the scope of this PR. If it can be done, we should address that I might've missed that possibility

rsolari commented:
I didn’t run the test myself, but I’m afk for the weekend. I’ll check when
I get back to a computer on Monday.

On Thu, Aug 15, 2019 at 4:53 PM Matthew Zipkin <notifications@github.com>
wrote:

> @rsolari <https://github.com/rsolari> were you able to OPEN a closed
> auction with a winner? That check should be happening already, out of the
> scope of this PR. If it can be done, we should address that I might've
> missed that possibility
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/handshake-org/hsd/pull/94?email_source=notifications&email_token=AACRVDE3BIZA3ZEVD67HJS3QEW65XA5CNFSM4GSPTIH2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD4M7A6Y#issuecomment-521793659>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AACRVDE5XS7NJFSQKQKOZL3QEW65XANCNFSM4GSPTIHQ>
> .
>

tynes commented on test/wallet-auction-test.js:
```diff
@@ -0,0 +1,191 @@
+  it('should mine enough blocks to expire auction', async () => {
+    for (
+      let i = 0;
+      i < network.names.biddingPeriod + network.names.revealPeriod;
```
Nit: precalculate `network.names.biddingPeriod + network.names.revealPeriod` into a variable as to not need to use multiple lines for the `for` statement

tynes commented on test/wallet-auction-test.js:
```diff
@@ -0,0 +1,191 @@
+    miner.addresses = [addr.getAddress().toString(network)];
+  });
+
+  it('should mine 20 blocks', async () => {
```
Is it possible to mine less blocks as to keep the test suite lighter? Could also move this out of an `it` and into `before`

tynes commented on test/wallet-auction-test.js:
```diff
@@ -0,0 +1,191 @@
+  });
+
+  it('should cleanup', async () => {
+    await miner.close();
```
This could go in the `after`

tynes commented on test/wallet-auction-test.js:
```diff
@@ -0,0 +1,191 @@
+  });
+
+  it('should add addrs to miner', async () => {
+    const addr = await winner.createReceive();
```
This could go in the `before`

pinheadmz commented:
@tynes thanks for the feedback, nits addressed and test is a lot cleaner. PR commits are cleaner. Rebased on master.

@rsolari although this PR doesn't test it explicitly (a fullnode would be required, ie integration test instead of unit test) I did run tests with the CLI using separate wallets and the behavior was as expected: unchanged. Namely, I got the error `Name is not available.` after an auction is closed with a winner, as well as any other state where a name is already `OPEN`.

rsolari commented:
@pinheadmz Thanks !

I also did some manual testing and correctly got `Name is not available` after an auction closed.

tynes commented on lib/wallet/txdb.js:
```diff
@@ -827,9 +827,20 @@ class TXDB {
+      if (hash) {
+        const names = this.wdb.network.names;
+        const period = names.biddingPeriod + names.revealPeriod;
+        const oldTX = await this.getTX(hash);
```
Nit: remove brackets for 1 line if statement, `oldTX` -> `tx` variable naming for minimal style

pinheadmz replied:
np removing brackets, thanks.
Don't wanna rename `oldTX` because variable `tx` is already defined as parameter to the function, its the transaction that _might_ be a double open.

tynes commented on lib/wallet/txdb.js:
```diff
@@ -827,9 +827,20 @@ class TXDB {
+        const oldTX = await this.getTX(hash);
+        if (oldTX.height !== -1 && oldTX.height + period < this.wdb.height) {
+          return false;
+        } else {
```
Nit: Unnecessary `else`

pinheadmz replied:
fixed.

pinheadmz commented:
@tynes nits addressed, rebased on master.

pinheadmz commented:
@boymanjor thanks dude, nits addressed, rebased on master to check test, and force pushed.
